File: server/routes/progress.ts
import express from "express";
import { Workout } from "../models/Workout";

const router = express.Router();

// Get volume progression data
router.get("/volume", async (req, res) => {
  try {
    const userId = "single"; // TODO: get from auth
    const { range = "last-30-days" } = req.query;

    // Calculate date range
    const endDate = new Date();
    let startDate = new Date();

    switch (range) {
      case "last-7-days":
        startDate.setDate(endDate.getDate() - 7);
        break;
      case "last-30-days":
        startDate.setDate(endDate.getDate() - 30);
        break;
      case "last-3-months":
        startDate.setMonth(endDate.getMonth() - 3);
        break;
      case "all-time":
        startDate = new Date(0); // Beginning of time
        break;
      default:
        startDate.setDate(endDate.getDate() - 30);
    }

    // Aggregate volume by date
    const volumeData = await Workout.aggregate([
      {
        $match: {
          userId,
          date: { $gte: startDate, $lte: endDate },
        },
      },
      {
        $group: {
          _id: {
            $dateToString: { format: "%Y-%m-%d", date: "$date" },
          },
          totalVolume: { $sum: "$metrics.totalVolume" },
          workoutCount: { $sum: 1 },
        },
      },
      {
        $sort: { _id: 1 },
      },
    ]);

    const formattedData = volumeData.map((item) => ({
      date: item._id,
      volume: item.totalVolume,
      workouts: item.workoutCount,
    }));

    res.json(formattedData);
  } catch (error) {
    console.error("Error fetching volume data:", error);
    res.status(500).json({ error: "Failed to fetch volume data" });
  }
});

// Get muscle group distribution
router.get("/muscle-distribution", async (req, res) => {
  try {
    const userId = "single";
    const { range = "last-30-days" } = req.query;

    // Calculate date range
    const endDate = new Date();
    let startDate = new Date();
    startDate.setDate(endDate.getDate() - 30); // Default to 30 days

    // This is simplified - you'd need to map exercises to muscle groups
    // For now, return mock data structure
    const muscleData = [
      { name: "Chest", value: 25 },
      { name: "Legs", value: 30 },
      { name: "Back", value: 20 },
      { name: "Shoulders", value: 15 },
      { name: "Arms", value: 10 },
    ];

    res.json(muscleData);
  } catch (error) {
    console.error("Error fetching muscle distribution:", error);
    res.status(500).json({ error: "Failed to fetch muscle distribution" });
  }
});

// Get quick stats
router.get("/stats", async (req, res) => {
  try {
    const userId = "single";
    const { range = "last-7-days" } = req.query;

    // Calculate date range
    const endDate = new Date();
    let startDate = new Date();
    startDate.setDate(endDate.getDate() - 7);

    // Get current period stats
    const currentStats = await Workout.aggregate([
      {
        $match: {
          userId,
          date: { $gte: startDate, $lte: endDate },
        },
      },
      {
        $group: {
          _id: null,
          totalVolume: { $sum: "$metrics.totalVolume" },
          totalSets: { $sum: "$metrics.numSets" },
          workoutCount: { $sum: 1 },
          avgWeight: { $avg: "$metrics.avgWeight" },
        },
      },
    ]);

    // Get previous period for comparison
    const prevStart = new Date(startDate);
    prevStart.setDate(prevStart.getDate() - 7);
    const prevStats = await Workout.aggregate([
      {
        $match: {
          userId,
          date: { $gte: prevStart, $lt: startDate },
        },
      },
      {
        $group: {
          _id: null,
          totalVolume: { $sum: "$metrics.totalVolume" },
          workoutCount: { $sum: 1 },
        },
      },
    ]);

    const current = currentStats[0] || {
      totalVolume: 0,
      workoutCount: 0,
      totalSets: 0,
      avgWeight: 0,
    };
    const previous = prevStats[0] || { totalVolume: 0, workoutCount: 0 };

    const volumeChange =
      previous.totalVolume > 0
        ? ((current.totalVolume - previous.totalVolume) /
            previous.totalVolume) *
          100
        : 0;

    const workoutChange =
      previous.workoutCount > 0
        ? ((current.workoutCount - previous.workoutCount) /
            previous.workoutCount) *
          100
        : 0;

    res.json({
      totalVolume: Math.round(current.totalVolume),
      workouts: current.workoutCount,
      avgWeight: Math.round(current.avgWeight),
      volumeChange: Math.round(volumeChange),
      workoutChange: Math.round(workoutChange),
    });
  } catch (error) {
    console.error("Error fetching stats:", error);
    res.status(500).json({ error: "Failed to fetch stats" });
  }
});

// Get strength trends per exercise
router.get("/strength-trends", async (req, res) => {
  try {
    const userId = "single";
    const { exerciseId, range = "last-30-days" } = req.query;

    const endDate = new Date();
    let startDate = new Date();
    startDate.setDate(endDate.getDate() - 30);

    const matchStage: any = {
      userId,
      date: { $gte: startDate, $lte: endDate },
    };

    // Unwind exercises array to work with individual exercises
    const trends = await Workout.aggregate([
      { $match: matchStage },
      { $unwind: "$exercises" },
      {
        $group: {
          _id: {
            exerciseId: "$exercises.exerciseId",
            date: { $dateToString: { format: "%Y-%m-%d", date: "$date" } },
          },
          maxWeight: { $max: "$exercises.sets.weight" },
          totalVolume: {
            $sum: {
              $reduce: {
                input: "$exercises.sets",
                initialValue: 0,
                in: {
                  $add: [
                    "$$value",
                    { $multiply: ["$$this.weight", "$$this.reps"] },
                  ],
                },
              },
            },
          },
          name: { $first: "$exercises.name" },
        },
      },
      { $sort: { "_id.date": 1 } },
    ]);

    // Group by exercise
    const groupedByExercise: any = {};
    trends.forEach((item) => {
      const exId = item._id.exerciseId;
      if (!groupedByExercise[exId]) {
        groupedByExercise[exId] = {
          exerciseId: exId,
          name: item.name,
          data: [],
        };
      }
      groupedByExercise[exId].data.push({
        date: item._id.date,
        maxWeight: item.maxWeight,
        volume: item.totalVolume,
      });
    });

    const result = Object.values(groupedByExercise);
    res.json(result);
  } catch (error) {
    console.error("Error fetching strength trends:", error);
    res.status(500).json({ error: "Failed to fetch strength trends" });
  }
});

// Get PR tracking and detection
router.get("/prs", async (req, res) => {
  try {
    const userId = "single";

    // Get all workouts and find PRs (max weight per exercise)
    const prs = await Workout.aggregate([
      { $match: { userId } },
      { $unwind: "$exercises" },
      { $unwind: "$exercises.sets" },
      {
        $group: {
          _id: "$exercises.exerciseId",
          name: { $first: "$exercises.name" },
          maxWeight: { $max: "$exercises.sets.weight" },
          maxWeightDate: {
            $max: {
              $cond: [
                {
                  $eq: [
                    "$exercises.sets.weight",
                    { $max: "$exercises.sets.weight" },
                  ],
                },
                "$date",
                null,
              ],
            },
          },
          totalSessions: { $sum: 1 },
        },
      },
      { $sort: { maxWeight: -1 } },
    ]);

    res.json(prs);
  } catch (error) {
    console.error("Error fetching PRs:", error);
    res.status(500).json({ error: "Failed to fetch PRs" });
  }
});

// Get workout frequency heatmap data
router.get("/heatmap", async (req, res) => {
  try {
    const userId = "single";
    const { range = "last-90-days" } = req.query;

    const endDate = new Date();
    let startDate = new Date();
    startDate.setDate(endDate.getDate() - 90);

    const workouts = await Workout.aggregate([
      {
        $match: {
          userId,
          date: { $gte: startDate, $lte: endDate },
        },
      },
      {
        $group: {
          _id: { $dateToString: { format: "%Y-%m-%d", date: "$date" } },
          count: { $sum: 1 },
          totalVolume: { $sum: "$metrics.totalVolume" },
        },
      },
      { $sort: { _id: 1 } },
    ]);

    const heatmapData = workouts.map((item) => ({
      date: item._id,
      count: item.count,
      volume: item.totalVolume,
      intensity: Math.min(item.count * 20, 100), // Scale intensity 0-100
    }));

    res.json(heatmapData);
  } catch (error) {
    console.error("Error fetching heatmap data:", error);
    res.status(500).json({ error: "Failed to fetch heatmap data" });
  }
});

// Get insights (plateaus, streaks, recommendations)
router.get("/insights", async (req, res) => {
  try {
    const userId = "single";
    const insights = [];

    // Get recent workouts
    const recentWorkouts = await Workout.find({ userId })
      .sort({ date: -1 })
      .limit(30);

    // Calculate streak
    let streak = 0;
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    for (let i = 0; i < recentWorkouts.length; i++) {
      const workoutDate = new Date(recentWorkouts[i].date);
      workoutDate.setHours(0, 0, 0, 0);
      const expectedDate = new Date(today);
      expectedDate.setDate(today.getDate() - i);

      if (workoutDate.getTime() === expectedDate.getTime()) {
        streak++;
      } else {
        break;
      }
    }

    if (streak >= 3) {
      insights.push({
        type: "success",
        title: "Amazing Streak!",
        message: `You've worked out ${streak} days in a row. Keep it up!`,
      });
    }

    // Check for plateaus (no volume increase in last 4 weeks)
    const fourWeeksAgo = new Date();
    fourWeeksAgo.setDate(fourWeeksAgo.getDate() - 28);

    const recentVolume = await Workout.aggregate([
      { $match: { userId, date: { $gte: fourWeeksAgo } } },
      { $group: { _id: null, avgVolume: { $avg: "$metrics.totalVolume" } } },
    ]);

    const olderVolume = await Workout.aggregate([
      {
        $match: {
          userId,
          date: {
            $lt: fourWeeksAgo,
            $gte: new Date(fourWeeksAgo.getTime() - 28 * 24 * 60 * 60 * 1000),
          },
        },
      },
      { $group: { _id: null, avgVolume: { $avg: "$metrics.totalVolume" } } },
    ]);

    if (recentVolume[0] && olderVolume[0]) {
      const volumeChange =
        ((recentVolume[0].avgVolume - olderVolume[0].avgVolume) /
          olderVolume[0].avgVolume) *
        100;

      if (volumeChange < 5) {
        insights.push({
          type: "warning",
          title: "Potential Plateau",
          message:
            "Your volume hasn't increased much in 4 weeks. Consider progressive overload.",
        });
      } else if (volumeChange > 15) {
        insights.push({
          type: "success",
          title: "Great Progress!",
          message: `Your volume is up ${volumeChange.toFixed(
            1
          )}% in the last 4 weeks!`,
        });
      }
    }

    res.json(insights);
  } catch (error) {
    console.error("Error fetching insights:", error);
    res.status(500).json({ error: "Failed to fetch insights" });
  }
});

export default router;

File: backend-go/routes/progress.go
package routes

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/sarthak0714/minmax-backend/db"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// RegisterProgressRoutes registers all progress routes
func RegisterProgressRoutes(router *gin.RouterGroup) {
	progress := router.Group("/progress")
	{
		progress.GET("/volume", getVolumeData)
		progress.GET("/stats", getStats)
		progress.GET("/strength-trends", getStrengthTrends)
		progress.GET("/prs", getPRs)
		progress.GET("/heatmap", getHeatmap)
		progress.GET("/insights", getInsights)
	}
}

// getVolumeData handles GET /api/progress/volume
func getVolumeData(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userID := "single" // TODO: get from auth
	rangeParam := c.DefaultQuery("range", "last-30-days")

	endDate := time.Now()
	var startDate time.Time
	var dateFormat string
	var groupByPeriod bool

	switch rangeParam {
	case "last-7-days":
		startDate = endDate.AddDate(0, 0, -7)
		dateFormat = "%Y-%m-%d"
		groupByPeriod = false
	case "last-30-days":
		startDate = endDate.AddDate(0, 0, -30)
		dateFormat = "%Y-%m-%d"
		groupByPeriod = false
	case "last-90-days":
		startDate = endDate.AddDate(0, 0, -90)
		dateFormat = "%Y-W%U" // Week format
		groupByPeriod = true
	case "last-180-days":
		startDate = endDate.AddDate(0, 0, -180)
		dateFormat = "%Y-W%U" // Week format
		groupByPeriod = true
	case "last-365-days":
		startDate = endDate.AddDate(0, 0, -365)
		dateFormat = "%Y-%m" // Month format
		groupByPeriod = true
	case "last-3-months":
		startDate = endDate.AddDate(0, -3, 0)
		dateFormat = "%Y-W%U"
		groupByPeriod = true
	case "all-time":
		startDate = time.Unix(0, 0)
		dateFormat = "%Y-%m" // Month format
		groupByPeriod = true
	default:
		startDate = endDate.AddDate(0, 0, -30)
		dateFormat = "%Y-%m-%d"
		groupByPeriod = false
	}

	collection := db.GetCollection("workouts")
	pipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"userId": userID,
			"date":   bson.M{"$gte": startDate, "$lte": endDate},
		}}},
		{{Key: "$group", Value: bson.M{
			"_id": bson.M{
				"$dateToString": bson.M{"format": dateFormat, "date": "$date"},
			},
			"totalVolume":  bson.M{"$sum": "$metrics.totalVolume"},
			"workoutCount": bson.M{"$sum": 1},
		}}},
		{{Key: "$sort", Value: bson.M{"_id": 1}}},
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch volume data"})
		return
	}
	defer cursor.Close(ctx)

	var results []bson.M
	if err := cursor.All(ctx, &results); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse volume data"})
		return
	}

	// For grouped data (weekly/monthly), just return aggregated results
	if groupByPeriod {
		formattedData := []gin.H{}
		for _, item := range results {
			formattedData = append(formattedData, gin.H{
				"date":     item["_id"].(string),
				"volume":   item["totalVolume"],
				"workouts": item["workoutCount"],
			})
		}
		c.JSON(http.StatusOK, formattedData)
		return
	}

	// For daily data, fill in all days in the range
	dataMap := make(map[string]bson.M)
	for _, item := range results {
		dateStr := item["_id"].(string)
		dataMap[dateStr] = item
	}

	formattedData := []gin.H{}
	currentDate := startDate
	for currentDate.Before(endDate) || currentDate.Equal(endDate) {
		dateStr := currentDate.Format("2006-01-02")
		if data, exists := dataMap[dateStr]; exists {
			formattedData = append(formattedData, gin.H{
				"date":     dateStr,
				"volume":   data["totalVolume"],
				"workouts": data["workoutCount"],
			})
		} else {
			// Fill with zeros for missing days
			formattedData = append(formattedData, gin.H{
				"date":     dateStr,
				"volume":   0,
				"workouts": 0,
			})
		}
		currentDate = currentDate.AddDate(0, 0, 1)
	}

	c.JSON(http.StatusOK, formattedData)
}

// getStats handles GET /api/progress/stats
func getStats(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userID := "single"
	rangeParam := c.DefaultQuery("range", "last-7-days")

	endDate := time.Now()
	var startDate time.Time
	var prevStartDate time.Time

	switch rangeParam {
	case "last-7-days":
		startDate = endDate.AddDate(0, 0, -7)
		prevStartDate = startDate.AddDate(0, 0, -7)
	case "last-30-days":
		startDate = endDate.AddDate(0, 0, -30)
		prevStartDate = startDate.AddDate(0, 0, -30)
	case "last-90-days":
		startDate = endDate.AddDate(0, 0, -90)
		prevStartDate = startDate.AddDate(0, 0, -90)
	case "last-180-days":
		startDate = endDate.AddDate(0, 0, -180)
		prevStartDate = startDate.AddDate(0, 0, -180)
	case "last-365-days":
		startDate = endDate.AddDate(0, 0, -365)
		prevStartDate = startDate.AddDate(0, 0, -365)
	case "all-time":
		startDate = time.Unix(0, 0)
		prevStartDate = time.Unix(0, 0)
	default:
		startDate = endDate.AddDate(0, 0, -7)
		prevStartDate = startDate.AddDate(0, 0, -7)
	}

	collection := db.GetCollection("workouts")

	// Get current period stats
	currentPipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"userId": userID,
			"date":   bson.M{"$gte": startDate, "$lte": endDate},
		}}},
		{{Key: "$group", Value: bson.M{
			"_id":          nil,
			"totalVolume":  bson.M{"$sum": "$metrics.totalVolume"},
			"totalSets":    bson.M{"$sum": "$metrics.numSets"},
			"workoutCount": bson.M{"$sum": 1},
		}}},
	}

	currentCursor, err := collection.Aggregate(ctx, currentPipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch stats"})
		return
	}
	defer currentCursor.Close(ctx)

	var currentStats []bson.M
	if err := currentCursor.All(ctx, &currentStats); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse stats"})
		return
	}

	// Get previous period for comparison
	prevPipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"userId": userID,
			"date":   bson.M{"$gte": prevStartDate, "$lt": startDate},
		}}},
		{{Key: "$group", Value: bson.M{
			"_id":          nil,
			"totalVolume":  bson.M{"$sum": "$metrics.totalVolume"},
			"workoutCount": bson.M{"$sum": 1},
		}}},
	}

	prevCursor, err := collection.Aggregate(ctx, prevPipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch previous stats"})
		return
	}
	defer prevCursor.Close(ctx)

	var prevStats []bson.M
	if err := prevCursor.All(ctx, &prevStats); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse previous stats"})
		return
	}

	current := gin.H{
		"totalVolume":  0.0,
		"workoutCount": 0,
		"totalSets":    0,
	}
	if len(currentStats) > 0 {
		current = gin.H(currentStats[0])
	}

	previous := gin.H{
		"totalVolume":  0.0,
		"workoutCount": 0,
	}
	if len(prevStats) > 0 {
		previous = gin.H(prevStats[0])
	}

	volumeChange := 0.0
	if prevVol, ok := previous["totalVolume"].(float64); ok && prevVol > 0 {
		if currVol, ok := current["totalVolume"].(float64); ok {
			volumeChange = ((currVol - prevVol) / prevVol) * 100
		}
	}

	workoutChange := 0.0
	if prevCount, ok := previous["workoutCount"].(int32); ok && prevCount > 0 {
		if currCount, ok := current["workoutCount"].(int32); ok {
			workoutChange = (float64(currCount-prevCount) / float64(prevCount)) * 100
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"totalVolume":   current["totalVolume"],
		"workouts":      current["workoutCount"],
		"volumeChange":  int(volumeChange),
		"workoutChange": int(workoutChange),
	})
}

// getStrengthTrends handles GET /api/progress/strength-trends
func getStrengthTrends(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userID := "single"
	rangeParam := c.DefaultQuery("range", "last-30-days")

	endDate := time.Now()
	var startDate time.Time

	switch rangeParam {
	case "last-7-days":
		startDate = endDate.AddDate(0, 0, -7)
	case "last-30-days":
		startDate = endDate.AddDate(0, 0, -30)
	case "last-90-days":
		startDate = endDate.AddDate(0, 0, -90)
	case "last-180-days":
		startDate = endDate.AddDate(0, 0, -180)
	case "last-365-days":
		startDate = endDate.AddDate(0, 0, -365)
	case "all-time":
		startDate = time.Unix(0, 0)
	default:
		startDate = endDate.AddDate(0, 0, -30)
	}

	collection := db.GetCollection("workouts")
	pipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"userId": userID,
			"date":   bson.M{"$gte": startDate, "$lte": endDate},
		}}},
		{{Key: "$unwind", Value: "$exercises"}},
		{{Key: "$group", Value: bson.M{
			"_id": bson.M{
				"exerciseId": "$exercises.exerciseId",
				"date":       bson.M{"$dateToString": bson.M{"format": "%Y-%m-%d", "date": "$date"}},
			},
			"maxWeight": bson.M{"$max": "$exercises.sets.weight"},
			"name":      bson.M{"$first": "$exercises.name"},
		}}},
		{{Key: "$sort", Value: bson.M{"_id.date": 1}}},
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch strength trends"})
		return
	}
	defer cursor.Close(ctx)

	var results []bson.M
	if err := cursor.All(ctx, &results); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse strength trends"})
		return
	}

	// Group by exercise
	groupedByExercise := make(map[string]gin.H)
	for _, item := range results {
		idMap := item["_id"].(bson.M)
		exID := idMap["exerciseId"].(primitive.ObjectID).Hex()

		if _, exists := groupedByExercise[exID]; !exists {
			groupedByExercise[exID] = gin.H{
				"exerciseId": exID,
				"name":       item["name"],
				"data":       []gin.H{},
			}
		}

		data := groupedByExercise[exID]["data"].([]gin.H)
		data = append(data, gin.H{
			"date":      idMap["date"],
			"maxWeight": item["maxWeight"],
		})
		groupedByExercise[exID]["data"] = data
	}

	result := make([]gin.H, 0, len(groupedByExercise))
	for _, v := range groupedByExercise {
		result = append(result, v)
	}

	c.JSON(http.StatusOK, result)
}

// getPRs handles GET /api/progress/prs
func getPRs(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userID := "single"
	collection := db.GetCollection("workouts")

	pipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{"userId": userID}}},
		{{Key: "$unwind", Value: "$exercises"}},
		{{Key: "$unwind", Value: "$exercises.sets"}},
		{{Key: "$group", Value: bson.M{
			"_id":           "$exercises.exerciseId",
			"name":          bson.M{"$first": "$exercises.name"},
			"maxWeight":     bson.M{"$max": "$exercises.sets.weight"},
			"totalSessions": bson.M{"$sum": 1},
		}}},
		{{Key: "$sort", Value: bson.M{"maxWeight": -1}}},
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch PRs"})
		return
	}
	defer cursor.Close(ctx)

	var results []bson.M
	if err := cursor.All(ctx, &results); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse PRs"})
		return
	}

	c.JSON(http.StatusOK, results)
}

// getHeatmap handles GET /api/progress/heatmap
func getHeatmap(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userID := "single"
	endDate := time.Now()
	startDate := endDate.AddDate(0, 0, -90)

	collection := db.GetCollection("workouts")
	pipeline := mongo.Pipeline{
		{{Key: "$match", Value: bson.M{
			"userId": userID,
			"date":   bson.M{"$gte": startDate, "$lte": endDate},
		}}},
		{{Key: "$group", Value: bson.M{
			"_id":         bson.M{"$dateToString": bson.M{"format": "%Y-%m-%d", "date": "$date"}},
			"count":       bson.M{"$sum": 1},
			"totalVolume": bson.M{"$sum": "$metrics.totalVolume"},
		}}},
		{{Key: "$sort", Value: bson.M{"_id": 1}}},
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch heatmap data"})
		return
	}
	defer cursor.Close(ctx)

	var results []bson.M
	if err := cursor.All(ctx, &results); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse heatmap data"})
		return
	}

	heatmapData := make([]gin.H, len(results))
	for i, item := range results {
		count := item["count"].(int32)
		intensity := count * 20
		if intensity > 100 {
			intensity = 100
		}

		heatmapData[i] = gin.H{
			"date":      item["_id"],
			"count":     count,
			"volume":    item["totalVolume"],
			"intensity": intensity,
		}
	}

	c.JSON(http.StatusOK, heatmapData)
}

// getInsights handles GET /api/progress/insights
func getInsights(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	userID := "single"
	insights := []gin.H{}

	collection := db.GetCollection("workouts")

	// Get recent workouts
	opts := options.Find()
	opts.SetSort(bson.M{"date": -1})
	opts.SetLimit(30)

	cursor, err := collection.Find(ctx, bson.M{"userId": userID}, opts)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch insights"})
		return
	}
	defer cursor.Close(ctx)

	var recentWorkouts []bson.M
	if err := cursor.All(ctx, &recentWorkouts); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse workouts"})
		return
	}

	// Calculate streak
	streak := 0
	today := time.Now()
	today = time.Date(today.Year(), today.Month(), today.Day(), 0, 0, 0, 0, today.Location())

	for i := 0; i < len(recentWorkouts); i++ {
		workoutDate := recentWorkouts[i]["date"].(primitive.DateTime).Time()
		workoutDate = time.Date(workoutDate.Year(), workoutDate.Month(), workoutDate.Day(), 0, 0, 0, 0, workoutDate.Location())
		expectedDate := today.AddDate(0, 0, -i)

		if workoutDate.Equal(expectedDate) {
			streak++
		} else {
			break
		}
	}

	if streak >= 3 {
		insights = append(insights, gin.H{
			"type":    "success",
			"title":   "Amazing Streak!",
			"message": fmt.Sprintf("You've worked out %d days in a row. Keep it up!", streak),
		})
	}

	c.JSON(http.StatusOK, insights)
}
